### ISSUE TITLE
Flatten example for strings does not work

### ISSUE THREAD

**Issue by sknop** [mrocklin/multipledispatch#52]
I tried the example in the documentation for flatten, which works for the first examples but fails on string:

RuntimeError: maximum recursion depth exceeded while calling a Python object

The problem is, I guess, that string is a subclass of Iterable: 

In [27]: isinstance("string", Iterable)
Out[27]: True

When iterating through a string, each element is again a string:
In [26]: for c in "string":
   ....:     print("{} : {}".format(c,c.__class__))
   ....:     
s : <class 'str'>
t : <class 'str'>
r : <class 'str'>
i : <class 'str'>
n : <class 'str'>
g : <class 'str'>

This creates an endless loop in the dispatch.

Has the example not been tested, or do I do something wrong?

**Comment by mrocklin** [comment]
Yes, that's a bug.  No, it's not yet being tested.  Can I interest you in submitting a fix?

**Comment by sknop** [comment]
might have to wait a few days, but I will make an attempt

while we are at the point of improvements, have you thought about ordering the different implementations hierarchically and providing a #next-method? This is what I used in Dylan (an advanced albeit now dead language featuring multi-methods similar to multiple dispatch).

next-method would allow you to write specialisations for a subclass and then call the original method. Very powerful concept

Cheers

Sven Erik

**Comment by mrocklin** [comment]
I'm roughly familiar with Dylan.  Clojure developers I know chat about it from time to time.

Ideas like next-method have come up.  I agree that they'd be powerful.  I have no immediate plans to extend `multipledispatch` to include them at the moment but am not against the idea.
